### DISCUSSION TITLE
Brickrail - Automatic Lego train control software

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#488] [Show and tell]
Hey everyone,

Today I'd like to share the first milestone of my software project using Pybricks: Brickrail. Today I first got automatic navigation between blocks working on a layout that incorporates a motorized switch. Here is a video recording this first success:
https://www.youtube.com/watch?v=xMpAScjcnIM

The software is built in the Godot game engine for its GUI features and real time graphics. It interfaces with a python script that uses `pybricksdev` to communicate with the Lego trains and layout controllers that run pybricks programs. Source code is published on my Github repository, but be aware that it is kind of a mess right now: https://github.com/Novakasa/brickrail

The full project is built to scale to an arbitrary track layout and multiple trains, stationary hubs and switches, and it has proper path finding to find the proper switch positions to successfully navigate the trains around the layout. For now though, I am glad it is working for one train featuring a single switch.

I will continue to work on this and I will probably produce some videos with multiple trains and switches soon.

**Comment by dlech** [comment]
Looking great!

**Comment by Novakasa** [comment]
Here's another recording that features 2 trains, a more complex layout and the ability to plan a route multiple blocks ahead:
https://www.youtube.com/watch?v=G22M5xvRr54
There's also a crash at the end since conflicting routes are currently still possible.
